influxql/functions.go

```diff
@@ -309,3 +316,322 @@ func (r *IntegerMovingAverageReducer) Emit() []FloatPoint {
 		},
 	}
 }
+
+// FloatHoltWintersReducer forecasts a series into the future.
+// This is done using the Holt-Winters damped method.
+//    1. Using the series the initial values are calculated using a SSE.
+//    2. The series is forecasted into the future using the iterative relations.
+type FloatHoltWintersReducer struct {
+	// Smoothing parameters
+	alpha,
+	beta,
+	gamma float64
+
+	// Dampening parameter
+	phi float64
+
+	// Season period
+	m        int
+	seasonal bool
+
+	// Horizon
+	h int
+
+	// Interval between points
+	interval int64
+
+	// Whether to include all data or only future values
+	includeFitData bool
+
+	// NelderMead optimizer
+	optim *neldermead.Optimizer
+	// Small difference bound for the optimizer
+	epsilon float64
+
+	y      []float64
+	points []FloatPoint
+}
+
+const (
+	defaultAlpha   = 0.5
+	defaultBeta    = 0.5
+	defaultGamma   = 0.5
+	defaultPhi     = 0.5
+	defaultEpsilon = 1.0e-4
+)
+
+// NewFloatHoltWintersReducer creates a new FloatHoltWintersReducer.
+func NewFloatHoltWintersReducer(h, m int, includeFitData bool, interval time.Duration) *FloatHoltWintersReducer {
+	seasonal := true
+	if m < 2 {
+		seasonal = false
+	}
+	return &FloatHoltWintersReducer{
+		alpha:          defaultAlpha,
+		beta:           defaultBeta,
+		gamma:          defaultGamma,
+		phi:            defaultPhi,
+		h:              h,
+		m:              m,
+		seasonal:       seasonal,
+		includeFitData: includeFitData,
+		interval:       int64(interval),
+		optim:          neldermead.New(),
+		epsilon:        defaultEpsilon,
+	}
+}
+
+func (r *FloatHoltWintersReducer) aggregate(time int64, value float64) {
+	r.points = append(r.points, FloatPoint{
+		Time:  time,
+		Value: value,
+	})
+}
+
+// AggregateFloat aggregates a point into the reducer and updates the current window.
+func (r *FloatHoltWintersReducer) AggregateFloat(p *FloatPoint) {
+	r.aggregate(p.Time, p.Value)
+}
+
+// AggregateInteger aggregates a point into the reducer and updates the current window.
+func (r *FloatHoltWintersReducer) AggregateInteger(p *IntegerPoint) {
+	r.aggregate(p.Time, float64(p.Value))
+}
+
+func (r *FloatHoltWintersReducer) roundTime(t int64) int64 {
+	return r.interval * ((t + r.interval/2) / r.interval)
+}
+
+func (r *FloatHoltWintersReducer) Emit() []FloatPoint {
+	if l := len(r.points); l < 2 || r.seasonal && l < r.m {
+		return nil
+	}
+	// First fill in r.y with values and NaNs for missing values
+	start, stop := r.roundTime(r.points[0].Time), r.roundTime(r.points[len(r.points)-1].Time)
+	count := (stop - start) / r.interval
+	r.y = make([]float64, 1, count)
```

Comment:
If `count` is zero, then this line will panic, because capacity cannot be smaller than length